src/task.ts

```diff
@@ -0,0 +1,450 @@
+                // The result is a list of all sub-task's results
+                const results = this.children.map((c) => c.result);
+                this.setValue(false, results);
+            }
+        } else {
+            // If any task failed, we fail the entire compound task
+            if (this.firstError === undefined) {
+                this.firstError = child.result as Error;
+                this.setValue(true, this.firstError);
+            }
+        }
+    }
+}
+
+/**
+ * @hidden
+ *
+ * A WhenAny task.
+ */
+export class WhenAnyTask extends CompoundTask {
+    /**
+     * @hidden
+     * Attempts to set a value to this task, given a completed sub-task
+     *
+     * @param child
+     *  The sub-task that just completed
+     */
+    public trySetValue(child: TaskBase): void {
+        // For a Task to have isError = true, it needs to contain within an Exception/Error datatype.
+        // However, WhenAny only contains Tasks as its result, so technically it "never errors out".
+        // The isError flag is used simply to determine if the result of the task should be fed in
+        // as a value, or as a raised exception to the generator code. For WhenAny, we always feed
+        // in the result as a value.
+        if (this.state === TaskState.Running) {
+            this.setValue(false, child);
+        }
+    }
+}
+
+/**
+ * @hidden
+ *
+ * A `-WithRetry` Task.
+ * It is modeled after a `WhenAllTask` because it decomposes
+ * into several sub-tasks (a growing sequence of timers and atomic tasks)
+ * that all need to complete before this task reaches an end-value.
+ */
+export class RetryableTask extends WhenAllTask {
+    private isWaitingOnTimer: boolean;
+    private numAttempts: number;
+    private error: any;
+
+    /**
+     * @hidden
+     * Construct a retriable task.
+     *
+     * @param innerTask
+     *  The task representing the work to retry
+     * @param retryOptions
+     *  The retrying settings
+     * @param executor
+     *  The taskOrchestrationExecutor managing the replay,
+     *  we use to to scheduling new tasks (timers and retries)
+     */
+    constructor(
+        public innerTask: DFTask,
+        private retryOptions: RetryOptions,
+        private executor: TaskOrchestrationExecutor
+    ) {
+        super([innerTask], innerTask.actionObj);
+        this.numAttempts = 1;
+        this.isWaitingOnTimer = false;
+    }
+
+    /**
+     * @hidden
+     * Attempts to set a value to this task, given a completed sub-task
+     *
+     * @param child
+     *  The sub-task that just completed
+     */
+    public trySetValue(child: TaskBase): void {
+        if (this.isWaitingOnTimer) {
+            this.isWaitingOnTimer = false;
+
+            // If we're out of retry attempts, we can set the output value
+            // of this task to be that of the last error we encountered
+            if (this.numAttempts >= this.retryOptions.maxNumberOfAttempts) {
```

Comment:
My previous comment seems to have gotten lost in the refactor. I think `>=` here needs to be `>`.

Reply:
I had responded to your previous comment, but as you point out: the refactor makes that thread hard to find :)

Here's the response I wrote to your original comment:

...
The ordering of statements is a bit confusing here, but I believe it's right.
Essentially, if we're on the 3rd attempt out of 3, and we're waiting on a timer, then we've already failed the 3rd attempt at re-scheduling the Task. In this state, we're just waiting for the timer to fire before deciding if we need to re-schedule the task again (and increase the attempt counter) or fail, as it's being done within this if-block.

Reply:
As a sanity check, I changed that check from `>=` to `>` and various legacy tests failed, which gives me some confidence that this logic is right. But I'm open to being challenged on this!

Reply:
Oh right, we did discuss this elsewhere. I really don't like the GitHub web UI for doing PRs - it's hard to keep track of feedback. 😣

Maybe I'm getting confused because there's actually two state variables at play, `isWaitingOnTimer` and `numAttempts`. The semantic meaning of `numAttempts` seems to change depending on the value of `isWaitingOnTimer`, which is really confusing. From a pure semantic standpoint, it's not correct to give up when attempt# == maxAttempts, thus I think most people who read this (and didn't also write the code) will be immediately confused. I wish it could be written in a way that was a bit more intuitive, but I trust you when you say it's correct and don't want to block on this.

Reply:
Thanks! I do want to make the code clear so I'll give it another go. My latest does a tiny refactor that should clarify the logic here. Please let me know what you think!